### ISSUE TITLE
Improve code exportability

### ISSUE THREAD

**Issue by budimanjojo** [siderolabs/talos#7777]
## Feature Request
Hello, I'm the maintainer of [a tool to generate Talos machineconfig files](https://github.com/budimanjojo/talhelper).
I use a lot of the code from here to do stuffs, so I'm very thankful for the beautifully written and opensource code :heart:.

I have just refactor the validation of the user input file to use more functions available from here and I have to rewrite this chunk of codes to validate `[]v1alpha1.Device`:
https://github.com/siderolabs/talos/blob/8670450d28040f35e08aa4d771a1415cd5c1920d/pkg/machinery/config/types/v1alpha1/v1alpha1_validation.go#L150-L195

It would be nice if those code can be wrapped with an exported function like `ValidateNetworkDevices(devices []*Device) ([]string, error)` but I see there's already such function in that file and it's just being called inside that chunk of codes in a for loop for each `*Device`. Maybe update that function or create a new function or rename that function?

I can create a PR if possible but I need instruction on how and whether to refactor it.

**Comment by smira** [comment]
Yep, this code is hard to follow, and hard to call with incomplete machine configuration. Are you looking to validate a single `Device` or a set of `Device`s? As there are some single-device checks and some which apply to a set of devices?

**Comment by smira** [comment]
Can you buiild a machine config and validate it as a whole? Does it make more sense or not?

I was thinking about adding a validation mode which doesn't complain about missing value, but only about values being set, which would work to validate e.g. config patches.

**Comment by budimanjojo** [comment]
@smira I do validate using the `Validate()` call after the machine config is being built. But I have a validation beforehand to validate the user provided config for my program too because there are required options so the program can behave correctly. So yeah currently it's being validated twice which is not cool but I prefer catching all misconfigured config before and after that the generated files should always pass the second validation so it's just a safeguard.

It's okay, let's see how your planned code turns out to be first then. Thank you 😃

**Comment by smira** [comment]
I mean support for validating partial machine config, something like that:

```yaml
machine:
  network:
    interfaces:
     - interface: eth0
       address: x.y.z.p
```

Right now it gives you back errors about required fields not being set in other parts of the machine config, but if that was fixed, I guess it would work great

**Comment by budimanjojo** [comment]
Oof that's not usable at my end because what I'm doing is the user provided something like this:
```yaml
nodes:
  - networkInterfaces:
      - interface: eth0
        address: x.y.z.p
    ...
```
And then I generate a machineconfig for each node.
Here's the example output of my validation, using the available code from here:
![Screenshot from 2023-09-22 16-30-09](https://github.com/siderolabs/talos/assets/13085918/31ae330a-818e-4be0-a360-cdc702f8788d)
